src/main/java/robot/subsystems/drivetrain/pure_pursuit/Path.java

```diff
@@ -154,4 +172,140 @@ public Path copy() {
+            //Creates a vector with the slope of the two points we are checking
+            //Sets the vectors magnitude to the constant of the spacing
+            //calculates the amount of fill-points that can fit between the two points.
+            pathVectors[i] = new Vector(this.getWaypoint(i), this.getWaypoint(i + 1));
+            AmountOfPoints = (int) Math.ceil(pathVectors[i].magnitude() / Constants.SPACING_BETWEEN_WAYPOINTS);
+            pathVectors[i] = pathVectors[i].normalize().multiply(Constants.SPACING_BETWEEN_WAYPOINTS);
+            for (int j = 0; j < AmountOfPoints; j++) {
+                newPoints.appendWaypoint(pathVectors[i].multiply(j).add(this.getWaypoint(i)));
+            }
+        }
+        return newPoints;
+
+
+    }
+
+    /**
+     * @param weight_data   amount of data
+     * @param weight_smooth amount of smooth
+     * @param tolerance     the min change between points
+     * @return the new path with the way points
+     * @author Paulo
+     * @author Lior
+     */
+    public Path generateSmoothing(double weight_data, double weight_smooth, double tolerance) {
+        Path newPathClass = this.copy();
+        double[][] newPath = new double[this.length()][2];
+        double a = weight_data;
+        double b = weight_smooth;
+        for (int i = 0; i < this.length(); i++) {
+            newPath[i][0] = this.getWaypoint(i).getX();
+            newPath[i][1] = this.getWaypoint(i).getY();
+
+        }
+        double[][] path = doubleArrayCopy(newPath);
+        double change = tolerance;
+        while (change >= tolerance) {
+            change = 0.0;
+            for (int i = 1; i < path.length - 1; i++)
+                for (int j = 0; j < path[i].length; j++) {
+                    double aux = newPath[i][j];
+                    newPath[i][j] += a * (path[i][j] - newPath[i][j]) + b *
+                            (newPath[i - 1][j] + newPath[i + 1][j] - (2.0 * newPath[i][j]));
+                    change += Math.abs(aux - newPath[i][j]);
+                }
+        }
+        for (int i = 0; i < this.length(); i++) {
+            Waypoint p = newPathClass.getWaypoint(i);
+            p.setX(newPath[i][0]);
+            p.setY(newPath[i][1]);
+            newPathClass.setWaypoint(i, p);
+        }
+        return newPathClass;
+    }
+
+    /**
+     * Attributes to all points their distance from the start.
+     */
+    private void generateDistance() {
+        this.recursiveDistance(this.length() - 1);
+    }
+
+    /**
+     * Returns the size of the largest length in the list.
+     *
+     * @param i index of current point
+     * @return returns sum of all distances before this point.
+     */
+    private double recursiveDistance(int i) {
+        if (i == 0)
+            return 0;
+        double d = recursiveDistance(i - 1) + Point.distance(path.get(i), path.get(i - 1));
+        Waypoint p = path.get(i);
+        p.setDistance(d);
+        path.set(i, p);
```

Comment:
this line is redundant. path.get(i) returns the same pointer, there is no need to save the changes to the waypoint back onto the path.
```suggestion
```